### PR TITLE
Port avg_pool3d() to ATen

### PR DESCRIPTION
**Pull request by skrah**
This will need a conflict resolution once avg_pool2d() has been merged.

### CODE REVIEW

**Comment by skrah** [comment]
@ezyang I think this one is ready now.  I've kept the existing frame functions but did not create new ones.

The existing `THCUNN` shape check omitted some trivial checks from the `THNN` one, so the unified checks are more restrictive now (same for CPU/GPU).

Also, `VolumetricAveragePooling.cu` did not check for "all same gpu".  I wonder if that was an omission, so I added that.

**Comment by ezyang** [comment]
> Also, VolumetricAveragePooling.cu did not check for "all same gpu". I wonder if that was an omission, so I added that.

Sounds like an omission; nice catch :)

File: aten/src/ATen/native/AveragePool3d.cpp [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+  IntArrayRef padding,
+  bool ceil_mode,
+  bool count_include_pad)
+{
+  // #20866 [JIT] stride.empty() is passed through
+  // #20866 [LIBTORCH] IntegrationTest.MNIST: padding.size() == 1
+  TORCH_INTERNAL_ASSERT(kernel_size.size() == 3 &&
+                        (stride.empty() || stride.size() == 3) &&
+                        (padding.size() == 1 || padding.size() == 3),
+    "avg_pool3d: all IntArrayRef sizes must be 3");
```
Comment (ezyang) [inline aten/src/ATen/native/AveragePool3d.cpp]: Huh, do we actually check this from the C++ frontend? If not, this probably should be a CHECK.

**Review by ezyang — APPROVED** [review]
Nifty!

**Comment by umanwizard** [comment]
This couldn't be landed due to internal issues. I am taking a look.

File: aten/src/ATen/native/AveragePool3d.cpp (hunk 1 above)
Comment (skrah) [reply aten/src/ATen/native/AveragePool3d.cpp]: @ezyang Good question, it actually came up now in #22032, which showed that people are using the C++ frontend like the Python functions (expecting that single integers are repeated automatically).

So it's a CHECK now (#22075).

**Comment by ezyang** [comment]
Oof, we should really fix this in the C++ frontend somehow.  I guess we
can bandaid it for now by adding logic to replicate in each operator
individually...

Excerpts from Stefan Krah's message of 2019-06-21 11:54:45 -0700:
> skrah commented on this pull request.
> 
> > +
> +void avg_pool3d_out_cpu_template(
> +  Tensor& output,
> +  const Tensor& input_,
> +  IntArrayRef kernel_size,
> +  IntArrayRef stride,
> +  IntArrayRef padding,
> +  bool ceil_mode,
> +  bool count_include_pad)
> +{
> +  // #20866 [JIT] stride.empty() is passed through
> +  // #20866 [LIBTORCH] IntegrationTest.MNIST: padding.size() == 1
> +  TORCH_INTERNAL_ASSERT(kernel_size.size() == 3 &&
> +                        (stride.empty() || stride.size() == 3) &&
> +                        (padding.size() == 1 || padding.size() == 3),
> +    "avg_pool3d: all IntArrayRef sizes must be 3");
> 
> @ezyang Good question, it actually came up now in #22032, which showed that people are using the C++ frontend like the Python functions (expecting that single integers are repeated automatically).
> 
> So it's a CHECK now (#22075).
>

**Pull request closed without merging**